
Status: Fixed in upstream CVS (gcc pre4.0).  Need to regenerate configure.

diff -ur mozilla/../libffi.old/configure.in mozilla/../libffi/configure.in
--- mozilla/../libffi.old/configure.in	2001-10-03 21:14:27.000000000 +0200
+++ mozilla/../libffi/configure.in	2004-07-22 03:00:24.000000000 +0200
@@ -43,7 +43,7 @@
 i*86-*-linux*) TARGET=X86; TARGETDIR=x86;;
 i*86-*-solaris*) TARGET=X86; TARGETDIR=x86;;
 i*86-*-beos*) TARGET=X86; TARGETDIR=x86;;
-i*86-*-freebsd*) TARGET=X86; TARGETDIR=x86;;
+i*86-*-freebsd* | i*86-*-kfreebsd*-gnu | i*86-*-knetbsd*-gnu) TARGET=X86; TARGETDIR=x86;;
 i*86-*-win32*) TARGET=X86_WIN32; TARGETDIR=x86;;
 i*86-*-cygwin*) TARGET=X86_WIN32; TARGETDIR=x86;;
 i*86-*-mingw*) TARGET=X86_WIN32; TARGETDIR=x86;;
diff -ur mozilla/../libffi.old/ltcf-c.sh mozilla/../libffi/ltcf-c.sh
--- mozilla/../libffi.old/ltcf-c.sh	2001-05-26 23:41:17.000000000 +0200
+++ mozilla/../libffi/ltcf-c.sh	2004-07-22 03:00:24.000000000 +0200
@@ -183,7 +183,7 @@
     whole_archive_flag_spec='-all_load $convenience'
     ;;
 
-  netbsd*)
+  netbsd* | knetbsd*-gnu)
     if echo __ELF__ | $CC -E - | grep __ELF__ >/dev/null; then
       archive_cmds='$LD -Bshareable $libobjs $deplibs $linker_flags -o $lib'
       wlarc=
@@ -403,7 +403,7 @@
     ;;
 
   # FreeBSD 3 and greater uses gcc -shared to do shared libraries.
-  freebsd*)
+  freebsd* | kfreebsd*-gnu)
     archive_cmds='$CC -shared -o $lib $libobjs $deplibs $compiler_flags'
     hardcode_libdir_flag_spec='-R$libdir'
     hardcode_direct=yes
@@ -434,7 +434,7 @@
     link_all_deplibs=yes
     ;;
 
-  netbsd*)
+  netbsd* | knetbsd*-gnu)
     if echo __ELF__ | $CC -E - | grep __ELF__ >/dev/null; then
       archive_cmds='$LD -Bshareable -o $lib $libobjs $deplibs $linker_flags'  # a.out
     else
diff -ur mozilla/../libffi.old/ltcf-cxx.sh mozilla/../libffi/ltcf-cxx.sh
--- mozilla/../libffi.old/ltcf-cxx.sh	2001-05-26 23:41:17.000000000 +0200
+++ mozilla/../libffi/ltcf-cxx.sh	2004-07-22 03:00:24.000000000 +0200
@@ -237,7 +237,7 @@
     # C++ shared libraries reported to be fairly broken before switch to ELF
     ld_shlibs=no
     ;;
-  freebsd*)
+  freebsd* | kfreebsd*-gnu)
     # FreeBSD 3 and later use GNU C++ and GNU ld with standard ELF
     # conventions
     ld_shlibs=yes
@@ -382,7 +382,7 @@
         ;;
     esac
     ;;
-  netbsd*)
+  netbsd* | knetbsd*-gnu)
     # NetBSD uses g++ - do we need to do anything?
     ;;
   osf3*)
@@ -737,7 +737,7 @@
           ;;
       esac
       ;;
-    freebsd*)
+    freebsd* | kfreebsd*-gnu)
       # FreeBSD uses GNU C++
       ;;
     hpux9* | hpux10* | hpux11*)
diff -ur mozilla/../libffi.old/ltcf-gcj.sh mozilla/../libffi/ltcf-gcj.sh
--- mozilla/../libffi.old/ltcf-gcj.sh	2001-05-26 23:41:17.000000000 +0200
+++ mozilla/../libffi/ltcf-gcj.sh	2004-07-22 03:00:24.000000000 +0200
@@ -176,7 +176,7 @@
       $CC $output_objdir/$soname-exp '$lt_cv_cc_dll_switch' -Wl,-e,'$dll_entry' -o $output_objdir/$soname '$ltdll_obj'$libobjs $deplibs $compiler_flags'
     ;;
 
-  netbsd*)
+  netbsd* | knetbsd*-gnu)
     if echo __ELF__ | $CC -E - | grep __ELF__ >/dev/null; then
       archive_cmds='$LD -Bshareable $libobjs $deplibs $linker_flags -o $lib'
       wlarc=
@@ -396,7 +396,7 @@
     ;;
 
   # FreeBSD 3 and greater uses gcc -shared to do shared libraries.
-  freebsd*)
+  freebsd* | kfreebsd*-gnu)
     archive_cmds='$CC -shared -o $lib $libobjs $deplibs $compiler_flags'
     hardcode_libdir_flag_spec='-R$libdir'
     hardcode_direct=yes
@@ -427,7 +427,7 @@
     link_all_deplibs=yes
     ;;
 
-  netbsd*)
+  netbsd* | knetbsd*-gnu)
     if echo __ELF__ | $CC -E - | grep __ELF__ >/dev/null; then
       archive_cmds='$LD -Bshareable -o $lib $libobjs $deplibs $linker_flags'  # a.out
     else
diff -ur mozilla/../libffi.old/ltconfig mozilla/../libffi/ltconfig
--- mozilla/../libffi.old/ltconfig	2001-05-26 23:41:17.000000000 +0200
+++ mozilla/../libffi/ltconfig	2004-07-22 03:00:24.000000000 +0200
@@ -1151,6 +1151,17 @@
   hardcode_into_libs=yes
   ;;
 
+kfreebsd*-gnu | knetbsd*-gnu)
+  version_type=linux
+  need_lib_prefix=no
+  need_version=no
+  library_names_spec='${libname}${release}.so$versuffix ${libname}${release}.so${major} ${libname}.so'
+  soname_spec='${libname}${release}.so$major'
+  shlibpath_var=LD_LIBRARY_PATH
+  shlibpath_overrides_runpath=yes
+  hardcode_into_libs=yes
+  ;;
+
 hpux9* | hpux10* | hpux11*)
   # Give a soname corresponding to the major version so that dld.sl refuses to
   # link against other versions.
